### ISSUE TITLE
What would be required to have this library perform the commissioning?

### ISSUE THREAD

**Issue by AlexDM0** [mfucci/node-matter#22]
Hi there! I'm currently working on a matter project using zephyr and the tooling available is quite poor to test with. It would be nice if we could do the commissioning from here as well. The android app, iOS app and the chip-device-ctrl tools are all kind of flaky.

I have a lot of experience working with BLE and have a nodejs lib that uses noble for ble. I might be able to help out tackling a part of this issue.

Any ideas on this?

**Comment by mfucci** [comment]
Hi Alex!

I am planning on adding commissioning over BLE / WiFi / Thread at some point, so I will be happy to help you with implementing this!
There is also on-going parallel work to build the Matter client.

BLE will introduce a platform dependency, since this is not supported natively by node.js, so the BLE platform-specific code should ideally be separated into his own project, like node-matter-ble-noble of something like that.

The work to implement this would be:
1 _ figure out how the BLE commissioning / communication works over BLE. My guess, scan for a specific advertisement instead of MDNS, use two attributes write / notify for the bi-directional communication, been careful how to chunk the messages so they stay below the MTU)
2 _ define an interface abstracting the BLE operations: node-matter will call the methods on the interface and node-matter-ble-noble will implement them
3 _ implement the BLE business logic in node-matter: this should not be too complex

I already started refactoring node-matter for supporting various types of io, with only io/udp supported yet (with no platform dependent implementation yet, but I need one since I want to support browsers and mobile apps).

BLE libs for node.js: you might want to double check which one would be the wider support / well maintained. Some wraps shell commands which is not great for many reasons, some integrate with DBUS to send low-level command, some have native code that call C libraries, a lot don't have typescript support, some are poorly maintained...

Let me know how you want to help! I also have some experience with BLE. It would help if I can have a test device similar to yours on my side to debug.

Thanks!
- Marco

**Comment by AlexDM0** [comment]
Hi Marco,

I don't have access to the specification (yet), so I could reverse engineer the chip device ctrl (https://github.com/project-chip/connectedhomeip/blob/master/src/controller/python/README.md). I'm not sure how stable this part of the spec. I have seen large changes between the nrf connect sdk 1.9.1 and 2.0.x regarding the required provisioning endpoints and clusters.

The BLE lib space is quite messy. We have settled on Noble since it supports windows, macOS and linux but it has been abandoned for quite a while. It does work well enough however, so and initial backend it could be fine.

I have not really read into the differences required to be a matter controller vs a matter peripheral (if any). For my use case, I'd like to be able to commission, to persist the fabric, get endpoints from devices in the fabric, read/write/get notifications from the clusters which support it.

I'm thinking if it makes sense to have everything in one lib or if it should be split like so (excuse the ble-ish terminology):
- matter central (provisioning, border router, etc..)
- matter peripheral (which is node-matter at the moment)
- matter core packets (consists of shared modules/endpoint definitions etc between central and peripheral)
- matter ble backend (can have multiple versions, depending on platform like noble or react-native)
- matter dashboard (connecting to matter central for UI interaction)

This way there can be a hierarchical build up of libraries and clear responsibilities for each library.

I could start by working on a matter central repo which can grow from the provisioning. I can take certain utils/package definitions from node-matter, which could later on be split out into a node-matter-core repo.

Finally, we can make a library which depends on all the parts to provide a fully featured api from a single import.

What do you think?

**Comment by mfucci** [comment]
Hi Alex,

**BLE libraries**: I did some research and I found (ordered by decreasing popularity):

- [noble](https://www.npmjs.com/package/noble) unmaintained for 4 years but maintained on [@abandonware](https://www.npmjs.com/package/@abandonware/noble). Supports OSX, Linux, Windows through native code (so required a working compiling env)
- [ble-host](https://www.npmjs.com/package/ble-host), maintained, use HCI, Linux only, no native code
- [node-ble](https://www.npmjs.com/package/node-ble), maintained, use bluez through D-Bus, Linux only, no native code
- [bluez](https://www.npmjs.com/package/bluez), maintained, use bluez through D-Bus, Linux only, no native code

Because of native code compilation, installing noble is far more complex than installing node-ble (for instance, I just tried to install both on my computer and only node-ble worked) so in my opinion a first implementation based on node-ble would be more popular. But you might have specific reasons to support Windows / OsX...

Other BLE implementations to consider in the future: Web Bluetooth Api for integration in a Chrome extension, iOs / Android webview support for integration in a mobile app.


To write the matter client, the easiest way is to write the integration test against the existing matter server: this will allow to test / fix very quickly.

Splitting into multiple repositories might make sense but I will be cautious about doing it too soon: this will add a lot of overhead, making changes harder to follow, versioning issues between repositories,  make refactoring one or two order of magnitude harder...
The main final benefit will be a reduction of the code size which can be achieved by other means. So not sure that it is worth it when the codebase is not stabilized. Let me know if you have other reasons in favor of splitting in repository now.
Even implementing BLE support should probably stays as a branch for some time (before API are stabilized) before being spited in its own repository

Data persistence is not implemented yet even for the server, a common solution should probably use for both.

Let me know if you want to tackle first the BLE support or the matter client. Some contributors were interested to work on the client as well and I can coordinate, help with the implementation.

Thanks!
- Marco

**Comment by Apollon77** [comment]
The OS support is a very good question ...
iobroker (which is one popular node.js based Open source platform) is is 90% Linux vs 9% Windows and "some" macOS and FreeBSD :-) (In Linux Docker is included)
For node-red there are no such statistic available ...

I also maintain two projects which uses BLE (homekit controller lib as example) and they also use noble ... because it was "the wy to go" as the libs started and also because of windows support. In fact yes noble is "somehow maintained in [@abandonware](https://www.npmjs.com/package/@abandonware/noble) but relies on community to deliver fixes and code changes, but yes then they can be released ... but it is no active development ... In fact C/C++ devs are needed that can fix such low level stuff if we run into issues ... so if we could have access to such guys (sorry I'm not really one of those) then boble is still a good solution to go tinto the future

Difficult decision ...

**Comment by AlexDM0** [comment]
Hi Marco,

I've only had bad experiences with dbus and bluez libraries (mismatches with the installed bluez version vs what the lib expected) on python and for us noble has worked out of the box on linux and macos :). Luck of the draw I suppose. We have had to use ble usb dongles in order for ble to work reliably, at least in ubuntu. I'm working on MacOS, due to React Native development on iOS. That limits a lot of the linux only libs if I don't want to boot up the VM.

With the existing Matter server, you're referring to the one being used by the [chip-device-ctrl](https://github.com/project-chip/connectedhomeip/tree/master/src/controller/python/chip/server)?

The main benefit of splitting up the repos is that it makes parallel development easier. Each part can be developed at its own pace and it can be combined later on. 

I'll look into the process of commissioning when I have some time. I unfortunately do not have a lot of time to work on this so I'll keep you posted!

Cheers

**Comment by mfucci** [comment]
@Apollon77 / @AlexDM0 : maybe the best path forward would be:

- since you have experience with noble, you write first an implementation using noble
- when it is working, I wrote another version based on [node-ble](https://www.npmjs.com/package/node-ble)
So the 90% Linux user won't have to deal with native code and MacOsX / Windows are still supported but with a more complex installation process.
What do you think?

@AlexDM0 : 
**Against the existing matter server**: the current node-matter codebase. I was planing on adding an integration test which is basically writing a matter client

**Splitting repo**: maybe we can start with using separate branches? As I was saying, the codebase will probably go through deep refactorings before being stable and more tooling for code styling will be added, so this will make is even harder if it is already split into multiple repos...

**Comment by Apollon77** [comment]
In fact for noble there is no real "more complex installation process" because it is all handled as normally by node-gyp and friends :-) And these days node.js without a basic build tooling do not make sense :-) But yes this would be a great way forward (and in fact e can still decide after noble version of we need another one)

**Comment by Apollon77** [comment]
PS: Beside "Multiple repos" we could also do a monorepo with lerna?

**Comment by mfucci** [comment]
**noble installation** : It is more complex... When I type:

- npm i node-ble, I get "added 17 packages, and audited 126 packages in 7s"
- npm i @abandonware/noble, I get "ERR! command failed: sh -c node-gyp rebuild"

Maybe the above issue is easy to fix but there are extra steps not done automatically.

**Multiple repos**: there will be a need for multi-repo for platform specific dependencies at some point.
I am just not sure this should be done now since it adds significant overhead. Building platform specific branch should be enough and when it works, the platform dependent piece can be split away in another repo.
I am not sure about the argument "it makes parallel development easier": using branches seems to me that it brings the same benefits and makes refactoring / merging / sharing tooling, code style, testing env / bug tracking easier...
Let me know if there is a benefit of multi-repos that I don't see for parallel development...

**Comment by mfucci** [comment]
Sorry, not much progress on my side: quite busy this week and I am still working on supporting attribute subscription (for the other open bug).
Next week, I will attend SiliconLabs Matter workshop so I am not sure how much time I will  have, but for sure the week after I will write a skeleton for the client side.

Let me know if you have already started something on your side.

**Comment by mfucci** [comment]
I did some refactoring to be able to create an integration test a Matter client: https://github.com/mfucci/node-matter/pull/26

- abstracted the UDP network interface -> BLE / Thread / browser-version interfaces can be implemented on top
- implemented a fake UDP network interface -> can be used in an integration test without doing any actual network operations
- introduced MatterClient class and refactored MatterServer to extract interfaces common with the client
- create the skeleton of the integration test: IntegrationTest.ts (obviously not passing right now...)

The next big refactoring to support a client would be to refactor protocols and clusters to extract common code and have a client and server implementation.

@Apollon77 / @AlexDM0 : let me know if you want to collaborate on that branch and I will invite you on the repo

**Comment by Apollon77** [comment]
For now I'm sorry but have other stuff on my plate to tackle first, but I will monitor.
With one of my planned stuff I most likely will do an ioBroker (smart home system) plugin that uses this lib to offer ioBroker switches via matter ... so a "real life usage" prototype ...

**Comment by AlexDM0** [comment]
I have a few things to finish up first. I'm not sure when I'll be able to get around to helping out, but I'll let you know when I'll be available. Best of luck with the lib!

**Comment by mfucci** [comment]
Ok, no problem!
I just finished writing about 70% of the commissioning: https://github.com/mfucci/node-matter/pull/26
It should be completed before the end of the week.

**Comment by Apollon77** [comment]
Wow, awesome work!

**Comment by Apollon77** [comment]
@mfucci does that mean that the library also has a "server" soon? Or do you plan to just have it for testing purposes?

**Comment by mfucci** [comment]
Not sure exactly what you refer to by a "server".

The library currently acts as a Matter device (server ?).

I am implementing a Matter controller (client ?). Right now, I am building it against the Matter device code in an integration test, but I will add a command line interface, probably similar to chip-tool. As this is meant to be a library, it will have API easy to use part of something else, like automating the control of multiple Matter devices.
I am planning on making it browser API compatible, so it can be used inside Chrome extension or inside a webview in a mobile app.

**Comment by mfucci** [comment]
Commissioning done.

After https://github.com/mfucci/node-matter/pull/26, controller -ip [DEVICE_IP] should perform the commission on a Matter fabric.
It is not a full fledged controller at this point, but the rest is now easy to implement.

The commissioning is done by calling the API (see Controller.ts as an example):
await client.commission(ip, port, discriminator, setupPin);

Feel free to leave comments on the PR, I will merge in one or two days.

**Comment by mfucci** [comment]
The code doing the commissioning is now merged into main so closing this issue.

@AlexDM0 @Apollon77
I invited you to join the repository. Past this major refactoring https://github.com/mfucci/node-matter/pull/33 , the code startucture starts to be stable enough to accept more help so let me know if you still want to help on BLE. I have already done initial work to abstract the communication channel type (BLE / WiFi / wired) and the platform implementation (Fake / Node / Node+ble-lib / browser).

Let's continue the discussion about BLE support in the forum
